tiledb/api/src/group/mod.rs

```diff
@@ -0,0 +1,559 @@
+        let mut vec_ptr: *const std::ffi::c_void = out_ptr!();
+        let mut vec_size: u32 = out_ptr!();
+        self.context.capi_return(unsafe {
+            ffi::tiledb_group_get_metadata_from_index(
+                c_context,
+                Self::capi(self),
+                index,
+                &mut key_ptr,
+                &mut key_len,
+                &mut c_datatype,
+                &mut vec_size,
+                &mut vec_ptr,
+            )
+        })?;
+
+        let c_key = unsafe { std::ffi::CStr::from_ptr(key_ptr) };
+        let key = c_key.to_string_lossy().into_owned();
+        let datatype = Datatype::try_from(c_datatype)?;
+        Ok(Metadata::new(key, datatype, vec_ptr, vec_size))
+    }
+
+    pub fn has_metadata_key<S>(
+        &self,
+        name: S,
+    ) -> TileDBResult<Option<Datatype>>
+    where
+        S: AsRef<str>,
+    {
+        let c_context = self.context.capi();
+        let c_name = cstring!(name.as_ref());
+        let mut c_datatype: ffi::tiledb_datatype_t = out_ptr!();
+        let mut exists: i32 = out_ptr!();
+        self.context.capi_return(unsafe {
+            ffi::tiledb_group_has_metadata_key(
+                c_context,
+                Self::capi(self),
+                c_name.as_ptr(),
+                &mut c_datatype,
+                &mut exists,
+            )
+        })?;
+        if exists == 0 {
+            return Ok(None);
+        }
+
+        let datatype = Datatype::try_from(c_datatype)?;
+        Ok(Some(datatype))
+    }
+
+    pub fn consolidate_metadata<S>(
+        &self,
+        config: &Config,
+        group_uri: S,
+    ) -> TileDBResult<()>
+    where
+        S: AsRef<str>,
+    {
+        let c_context = self.context.capi();
+        let c_group_uri = cstring!(group_uri.as_ref());
+        let cfg = config.capi();
+        self.context.capi_return(unsafe {
+            ffi::tiledb_group_consolidate_metadata(
+                c_context,
+                c_group_uri.as_ptr(),
+                cfg,
+            )
+        })?;
+        Ok(())
+    }
+
+    pub fn vacuum_metadata<S>(
+        &self,
+        config: &Config,
+        group_uri: S,
+    ) -> TileDBResult<()>
+    where
+        S: AsRef<str>,
+    {
+        let c_context = self.context.capi();
+        let c_group_uri = cstring!(group_uri.as_ref());
+        let cfg = config.capi();
+        self.context.capi_return(unsafe {
+            ffi::tiledb_group_vacuum_metadata(c_context, c_group_uri.as_ptr(), cfg)
+        })?;
+        Ok(())
+    }
+}
+
+impl Drop for Group<'_> {
+    fn drop(&mut self) {
+        let c_context = self.context.capi();
+        let c_group = Self::capi(self);
+
+        let mut c_open: i32 = out_ptr!();
+        self.context
+            .capi_return(unsafe {
+                ffi::tiledb_group_is_open(c_context, c_group, &mut c_open)
+            })
+            .expect("TileDB internal error when checking for open group.");
+        if c_open > 0 {
```

Comment:
In my opinion it should not be possible to construct a `Group` if the group is *not* open.

I'm fine with this being present as a failsafe mechanism *somewhere* but:
1) I think it should probably  be in `open`, to prevent constructing the `Group` (unless there are APIs which are valid to call on a closed group? is that the case?); and/or
2) it should be remarked upon with a comment that we don't really expect it ever to find a closed group.

Reply:
Wait no, the Group is constructed in create, and then opened once it is created. And the function open itself returns a Self, so by construction it is open? And close is called upon dropping an object.

Reply:
I should be clearer - a user of the API should not be able to get a Group type object unless the group it represents is open.  Your code currently does this, since `Group::open` returns `Err` if `tiledb_group_open` fails.  So the check here to `tiledb_group_is_open` in the destructor is kind of misleading, because it communicates that we might have a Group type object which represents a non-open group.

Unless the group can spontaneously close - I sure hope it can't!

I think you can add clarity to the use of `tiledb_group_is_open` by either calling it in `open` to prevent `open` from returning successfully, or by leaving a hard-to-miss comment here that we don't ever expect the group to be closed at this point.

Reply:
Fixed this!

Reply:
Oh, good point, you have a consuming method `delete_group` and then also this `Drop` impl.

Please leave a comment here though because it is not obvious why this is needed

Reply:
Wrote a comment!